db/background_mgr_attachment_migration.go

```diff
@@ -0,0 +1,351 @@
+			return true
+		}
+
+		// Don't want to process raw binary docs
+		// The binary check should suffice but for additional safety also check for empty bodies
+		if event.DataType == base.MemcachedDataTypeRaw || len(event.Value) == 0 {
+			return true
+		}
+
+		// We only want to process full docs. Not any sync docs.
+		if strings.HasPrefix(docID, base.SyncDocPrefix) {
+			return true
+		}
+
+		a.DocsProcessed.Add(1)
+		syncData, _, _, err := UnmarshalDocumentSyncDataFromFeed(event.Value, event.DataType, collection.userXattrKey(), false)
+		if err != nil {
+			failProcess(err, "error unmarshaling document %s: %v", base.UD(docID), err)
+		}
+
+		if syncData == nil || syncData.Attachments == nil {
+			// no attachments to migrate
+			return true
+		}
+
+		collCtx := collection.AddCollectionContext(ctx)
+		collWithUser := &DatabaseCollectionWithUser{
+			DatabaseCollection: collection,
+		}
+		// xattr migration to take place
+		err = collWithUser.MigrateAttachmentMetadata(collCtx, docID, event.Cas, syncData)
+		if err != nil {
+			failProcess(err, "error migrating document attachment metadata for doc: %s: %v", base.UD(docID), err)
+		} else {
+			a.DocsChanged.Add(1)
+		}
+		return true
+	}
+
+	bucket, err := base.AsGocbV2Bucket(db.Bucket)
+	if err != nil {
+		return err
+	}
+
+	currCollectionIDs := db.GetCollectionIDs()
+	checkpointPrefix := db.MetadataKeys.DCPCheckpointPrefix(db.Options.GroupID) + "att_migration:"
+
+	// check for mismatch in collection id's between current collections on the db and prev run
+	err = a.resetDCPMetadataIfNeeded(ctx, db.MetadataStore, db, checkpointPrefix, currCollectionIDs)
+	if err != nil {
+		return err
+	}
+
+	a.SetCollectionIDs(currCollectionIDs)
+	dcpOptions := getMigrationDCPClientOptions(currCollectionIDs, db.Options.GroupID, checkpointPrefix)
+	dcpFeedKey := GenerateAttachmentMigrationDCPStreamName(a.MigrationID)
+	dcpClient, err := base.NewDCPClient(ctx, dcpFeedKey, callback, *dcpOptions, bucket)
+	if err != nil {
+		base.WarnfCtx(ctx, "[%s] Failed to create attachment migration DCP client: %v", migrationLoggingID, err)
+		return err
+	}
+	base.InfofCtx(ctx, base.KeyAll, "[%s] Starting DCP feed %q for attachment migration", migrationLoggingID, dcpFeedKey)
+
+	doneChan, err := dcpClient.Start()
+	if err != nil {
+		base.WarnfCtx(ctx, "[%s] Failed to start attachment migration DCP feed: %v", migrationLoggingID, err)
+		_ = dcpClient.Close()
+		return err
+	}
+	base.DebugfCtx(ctx, base.KeyAll, "[%s] DCP client started for Attachment Migration.", migrationLoggingID)
+
+	select {
+	case <-doneChan:
+		base.InfofCtx(ctx, base.KeyAll, "[%s] Finished migrating attachment metadata from sync data to global sync data. %d/%d docs changed", migrationLoggingID, a.DocsChanged.Value(), a.DocsProcessed.Value())
+		err = dcpClient.Close()
+		if err != nil {
+			base.WarnfCtx(ctx, "[%s] Failed to close attachment migration DCP client! %v", migrationLoggingID, err)
+			return err
```

Comment:
I think this is a warning but I don't think this should fail the attachment process, I think that we should probably continue through this block.

(I don't think this is at all likely to fail)